### ISSUE TITLE
Stream semantics difference between CUDA/HIP

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#5163]
https://github.com/cupy/cupy/pull/5091#issuecomment-821946143
I think this needs to be documented in CuPy as users are almost impossible to know what's happening.
We may need to consider providing some remedies for this (e.g., wait until `hipStreamQuery` succeeds before destroying a stream when in HIP).

**Comment by leofang** [comment]
@amathews-amd Could you confirm `hipStreamDestroy()` has a different semantics when it's on AMD GPUs?

**Comment by amathews-amd** [comment]
I found that this might be the case. See
https://github.com/ROCmSoftwarePlatform/rocBLAS/pull/1171 

Paging @ekuznetsov139

**Comment by emcastillo** [comment]
Show a warning if there is work left before destroying? (by @kmaehashi )

**Comment by kmaehashi** [comment]
Thanks @amathews-amd! Understood that the semantics difference is by design.

> Documents that it's the user's responsibility to call hipStreamSynchronize() before calling rocblas_set_stream() and before calling hipStreamDestroy() when switching between non-default streams

Ideas what I came up with to mitigate the gap between CUDA/HIP:
1. Documentation
2. Call `hipStreamQuery` before calling `hipStreamDestroy` and show warning if the work is in progress
3. Call `hipStreamSynchronize` before calling `hipStreamDestroy`

I talked offline with other dev members and the idea 1 & 2 sounds reasonable.

**Comment by amathews-amd** [comment]
@leofang @kmaehashi Opened an internal ticket (SWDEV-285318) to recheck design of  hipStreamDestroy.

**Comment by kmaehashi** [comment]
Thank you @amathews-amd! I believe aligned behavior would be really helpful for other projects too.

**Comment by amathews-amd** [comment]
Hi @kmaehashi , 

The ROCclr team mentioned that the documentation might be out of date, and for all cases (except some corner cases), hipStreamDestroy will wait on work inflight. I tried to create a reproducer from your code to show the issue, but it looks like hipStreamDestroy on host thread is blocked till the queue is done.

Could you take a look at the reproducer below, and check if it matches your use-case, and run it on your system ? If you are able to reproduce the issue, please add the updated reproducer code, do 'rocminfo' to collect device info, and include the ROCm version.

Thanks,
Aswin. 

`
#include <stdio.h>
#include <iostream>
#include "hip/hip_runtime.h"
#include <ctime>

#ifndef ITERATIONS
#define ITERATIONS 1
#endif

__global__ void wait_and_write(long long *x, long long y) {
  clock_t start = clock();
  clock_t now;
  for (;;) {
    now = clock();
    clock_t cycles = now > start ? now - start : now + (0xffffffff - start);
    if (cycles >= 1000000000) {
      break;
    }
  }
  x[0] += y;
  x[1] = now;  // in case the compiler optimizing away the entire loop
}

int main()
{
        int device_count = 0;
        hipGetDeviceCount(&device_count);

        std::cout<< "Found " << device_count << " HIP devices." << std::endl;
        for( int i = 0; i < device_count; i++ )
        {
                std::cout << "Device " << i << std::endl;
                hipDeviceProp_t props;
                hipGetDeviceProperties(&props, i);
                std::cout << " Name: " << props.name << std::endl;
        }


        hipSetDevice(0);

        long long *A_d;
        long long *A_h;

        size_t N = 2;
        size_t Nbytes = N * sizeof( long long );

        // Allocating host memory
        A_h = (long long*) malloc(Nbytes);
        for( size_t i = 0; i < N ; i++ )
        {
                A_h[i] = 0;
        }

        // Allocating device memory
        hipMalloc(&A_d, Nbytes);

        for(int iter = 0; iter< ITERATIONS; iter++) {

                hipStream_t stream;
                hipStreamCreate(&stream);

                // Copy host to device
                hipMemcpyAsync(A_d, A_h, Nbytes, hipMemcpyHostToDevice, stream);

                // launch kernel
                const unsigned blocks = 1;
                const unsigned threadsPerBlock = 1;

                // Queue up kernel launches
                std::cout<< clock() << ": Calling kernel with y=1 " << std::endl;
                hipLaunchKernelGGL(wait_and_write, dim3(blocks), dim3(threadsPerBlock), 0, stream, A_d, 1);
                std::cout<< clock() << ": DONE Calling kernel " << std::endl;

                std::cout<< clock() << ": Calling kernel with y=2 " << std::endl;
                hipLaunchKernelGGL(wait_and_write, dim3(blocks), dim3(threadsPerBlock), 0, stream, A_d, 2);
                std::cout<< clock() << ": DONE Calling kernel " << std::endl;

                std::cout<< clock() << ": Calling kernel with y=4 " << std::endl;
                hipLaunchKernelGGL(wait_and_write, dim3(blocks), dim3(threadsPerBlock), 0, stream, A_d, 4);
                std::cout<< clock() << ": DONE Calling kernel " << std::endl;

                std::cout<< clock() << ": Calling kernel with y=4 " << std::endl;
                hipLaunchKernelGGL(wait_and_write, dim3(blocks), dim3(threadsPerBlock), 0, stream, A_d, 8);
                std::cout<< clock() << ": DONE Calling kernel " << std::endl;


                //hipStreamSynchronize(stream);

                std::cout<< clock() << ": StreamDestroy " << std::endl;
                hipStreamDestroy(stream);
                std::cout<< clock() << ": DONE StreamDestroy " << std::endl;

                // Copy device to host
                std::cout<< clock() << ": Copy Data out " << std::endl;
                hipMemcpy(A_h, A_d, Nbytes, hipMemcpyDeviceToHost);
                std::cout<< clock() << ": DONE Copy Data out " << std::endl;


                // Verify
                for( size_t i = 0; i < N; i++ )
                {
                        std::cout << A_h[i] << " ";
                }
                if( A_h[0] != 15 || A_h[1] == 0 )
                        std::cout << "FAILED" << std::endl;
                else
                        std::cout << "PASSED" << std::endl;
                std::cout << std::endl;
        }

        // free up host memory
        free( A_h );
}
`

**Comment by kmaehashi** [comment]
Thank you @amathews-amd! I tried the reproducer (changed `#define ITERATIONS 1` to `#define ITERATIONS 2`) in my local environment and it shows FAILED on the second iteration, using ROCm 4.2.0.

```
% hipcc test2.cc  
% ./a.out
Found 2 HIP devices.
Device 0
 Name: Vega 10 [Radeon Instinct MI25]
Device 1
 Name: Vega 10 XTX [Radeon Vega Frontier Edition]
353310: Calling kernel with y=1 
353992: DONE Calling kernel 
354002: Calling kernel with y=2 
354012: DONE Calling kernel 
354015: Calling kernel with y=4 
354019: DONE Calling kernel 
354022: Calling kernel with y=4 
354067: DONE Calling kernel 
354070: StreamDestroy 
3240886: DONE StreamDestroy 
3240934: Copy Data out 
3256489: DONE Copy Data out 
15 148931736197 PASSED

3268877: Calling kernel with y=1 
3268957: DONE Calling kernel 
3268984: Calling kernel with y=2 
3269002: DONE Calling kernel 
3269020: Calling kernel with y=4 
3269037: DONE Calling kernel 
3269053: Calling kernel with y=4 
3269068: DONE Calling kernel 
3269084: StreamDestroy 
6138320: DONE StreamDestroy 
6138337: Copy Data out 
6140804: DONE Copy Data out 
30 152929394033 FAILED
```

[rocminfo.txt](https://github.com/cupy/cupy/files/6470281/rocminfo.txt)

**Comment by amathews-amd** [comment]
Sorry, there is a bug in my code. 

You need to move 
`
for( size_t i = 0; i < N ; i++ )
        {
                A_h[i] = 0;
        }
`

inside the 
`
        for(int iter = 0; iter< ITERATIONS; iter++) {
`
loop.

**Comment by kmaehashi** [comment]
Ah ok, the FAIL disappeared now.

@takagi Do you remember whether the original test failures were flaky, or always reproducible?

**Comment by takagi** [comment]
Yes. It was always reproducible (with longer enough busy-wait).

**Comment by amathews-amd** [comment]
The ROCclr team is moving ahead with documentation update on design that the runtime wont destroy work in flight. 

Please let us know if you find anything to the contrary, and we can open a new ticket.

**Comment by amathews-amd** [comment]
@kmaehashi @takagi , by any chance, is this multithreaded, such that the thread that calls hipStreamDestroy() is being killed, when the work in queue is not yet completed ? Could you also run your original test that failed to root-cause the issue ?

**Comment by takagi** [comment]
No, the case I faced was single threaded. I will run the test again shortly with ROCm 4.2.0.

**Comment by takagi** [comment]
Hmm... the original test https://github.com/cupy/cupy/pull/5091 still fails with ROCm 4.2.0. I also checked the code you showed at https://github.com/cupy/cupy/issues/5163#issuecomment-840086050  and it passed on the same environment. It is supposed that I have something to investigate further...

**Comment by leofang** [comment]
> Hmm... the original test #5091 still fails with ROCm 4.2.0.

@takagi Oh? I thought we fixed it! Could you try if #5172 fixes #5091? #5172 now ensures a stream is not dead within its context.

**Comment by takagi** [comment]
Okay, let me try it. I will also check what the original test does again.

**Comment by leofang** [comment]
Though I think about it again, and not sure if it's not stream lifetime issue what else it could be...

**Comment by takagi** [comment]
I tried with #5172 with/without the change in #5091 (on ROCm 4.2.0). #5172 didn't make the original fail (before the change in #5091) away. And with both #5172 and #5091, I got an error the way I didn't ever see.

With #5172, with #5091
```
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference: 13744632839234567870
E       Max relative difference: inf
E        x: array([13744632839234567870, 13744632839234567870], dtype=uint64)
E        y: array([0, 0], dtype=uint64)
```

With #5172, without #5091
```
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference: 751925723207
E       Max relative difference: inf
E        x: array([           1, 751925723207], dtype=uint64)
E        y: array([0, 0], dtype=uint64)
```

Without #5172, with #5091
```
Passed
```

Without #5172, without #5091
```
E       Mismatched elements: 2 / 2 (100%)
E       Max absolute difference: 749637775449
E       Max relative difference: inf
E        x: array([           1, 749637775449], dtype=uint64)
E        y: array([0, 0], dtype=uint64)
```

**Comment by takagi** [comment]
> And with both #5172 and #5091, I got an error the way I didn't ever see.

Currently, in `ndarray.copy`, we change the active device to the one where src data is when we use HIP. If I modify the part to keep the active device to the one where dest data is as we do with CUDA, the test (with both #5172 and #5091) passes.

https://github.com/cupy/cupy/blob/4ed250d65ab2dc31a4c749f61a7ce9dc0a684ea8/cupy/_core/core.pyx#L513-L523

**Comment by takagi** [comment]
(Mistakenly closed)

**Comment by takagi** [comment]
> And with both #5172 and #5091, I got an error the way I didn't ever see.

Looks the way `hipMemcpyAsync` works is changed ("to set the current device ..." -> "to use a stream ...").

> For multi-gpu or peer-to-peer configurations, it is recommended to use a stream which is a attached to the device where the src data is physically located. 

https://rocm-developer-tools.github.io/HIP/group__Memory.html#gad55fa9f5980b711bc93c52820149ba18

**Comment by takagi** [comment]
Or, perhaps, the original documentation was wrong, and the way `hipMemcpyAsync` works does not change? @amathews-amd

**Comment by takagi** [comment]
I've found that `hipMemcpy` says "to set the current device ..." and `hipMemcpyAsync` says "to use a stream ...". I had to consider this in https://github.com/cupy/cupy/pull/5004 where I changed it from synchronous memcpy to asynchronous one.
